### ISSUE TITLE
sparse_categorical_crossentropy and temporal sample weight

### ISSUE THREAD

**Issue by nadavbra** [tensorflow/tensorflow#45966]
I am training a (BERT-like) language model with Keras, using _sparse_categorical_crossentropy_ loss. As with any training of language models, I have input sequences of different lengths, meaning that I need to add padding at the ends of most input sequences in order for them to fit a uniform batch size. When it comes to the outputs, however, I do not want to force the model to learn unnecessary padding tokens; rather, I'd just like it to ignore the outputs corresponding to positions where the input has a padding token.

The most sensible solution I was able to think of for doing that was using a _sample_weight_ with zeros at positions corresponding to padding in the input, and ones at all other (actual) positions. This requires setting _sample_weight_mode = 'temporal'_.

The problem is that _sparse_categorical_crossentropy_ doesn't seem to support that, and I get the following error:
_**ValueError: Found a sample_weight array for an input with shape (8108, 512). Timestep-wise sample weighting (use of sample_weight_mode="temporal") is restricted to outputs that are at least 3D, i.e. that have a time dimension.**_

Is there a way of using temporal sample weight with _sparse_categorical_crossentropy_  in Keras? 
Alternatively, is there another approach I should consider to achieve what I want, or any workaround I could resort to (for making Keras ignore positions with padding in the input during training)?

Thank you!

**Comment by amahendrakar** [comment]
@nadavbra,
In order to reproduce the issue reported here, could you please provide the TensorFlow version, the complete code and the dataset you are using. Thanks!

**Comment by nadavbra** [comment]
**I use TensorFlow 2.0.0.**

Here is a simple toy example reproducing the problem:

```python
import numpy as np
import tensorflow as tf
from tensorflow import keras

dataset_size = 3200
seq_len = 20
vocab_size = 10

X = np.random.randint(0, vocab_size, (dataset_size, seq_len))
Y = X

input_layer = keras.layers.Input(shape = (seq_len,), dtype = np.int32)
embedding_layer = keras.layers.Embedding(vocab_size, 5)(input_layer)
output_layer = keras.layers.Dense(vocab_size, activation = 'softmax')(embedding_layer)

model = keras.models.Model(inputs = input_layer, outputs = output_layer)
model.compile(loss = 'sparse_categorical_crossentropy', optimizer = keras.optimizers.Adam(), sample_weight_mode = 'temporal')
model.fit(X, Y)
```

Running this gives me the error:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-0ba618c51df7> in <module>
     16 model = keras.models.Model(inputs = input_layer, outputs = output_layer)
     17 model.compile(loss = 'sparse_categorical_crossentropy', optimizer = keras.optimizers.Adam(), sample_weight_mode = 'temporal')
---> 18 model.fit(X, Y)

/usr/local/tensorflow/avx-avx2-gpu/2.0.0/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py in fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)
    726         max_queue_size=max_queue_size,
    727         workers=workers,
--> 728         use_multiprocessing=use_multiprocessing)
    729 
    730   def evaluate(self,

/usr/local/tensorflow/avx-avx2-gpu/2.0.0/python3.7/site-packages/tensorflow_core/python/keras/engine/training_arrays.py in fit(self, model, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, **kwargs)
    640         steps=steps_per_epoch,
    641         validation_split=validation_split,
--> 642         shuffle=shuffle)
    643 
    644     if validation_data:

/usr/local/tensorflow/avx-avx2-gpu/2.0.0/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py in _standardize_user_data(self, x, y, sample_weight, class_weight, batch_size, check_steps, steps_name, steps, validation_split, shuffle, extract_tensors_from_dataset)
   2528           training_utils.standardize_weights(ref, sw, cw, mode)
   2529           for (ref, sw, cw, mode) in zip(y, sample_weights, class_weights,
-> 2530                                          feed_sample_weight_modes)
   2531       ]
   2532       # Check that all arrays have the same length.

/usr/local/tensorflow/avx-avx2-gpu/2.0.0/python3.7/site-packages/tensorflow_core/python/keras/engine/training.py in <listcomp>(.0)
   2527       sample_weights = [
   2528           training_utils.standardize_weights(ref, sw, cw, mode)
-> 2529           for (ref, sw, cw, mode) in zip(y, sample_weights, class_weights,
   2530                                          feed_sample_weight_modes)
   2531       ]

/usr/local/tensorflow/avx-avx2-gpu/2.0.0/python3.7/site-packages/tensorflow_core/python/keras/engine/training_utils.py in standardize_weights(y, sample_weight, class_weight, sample_weight_mode)
    884     if len(y.shape) < 3:
    885       raise ValueError('Found a sample_weight array for '
--> 886                        'an input with shape ' + str(y.shape) + '. '
    887                        'Timestep-wise sample weighting (use of '
    888                        'sample_weight_mode="temporal") is restricted to '

ValueError: Found a sample_weight array for an input with shape (3200, 20). Timestep-wise sample weighting (use of sample_weight_mode="temporal") is restricted to outputs that are at least 3D, i.e. that have a time dimension.
```

Note that everything works fine when removing the `sample_weight_mode = 'temporal'` option.

**Comment by amahendrakar** [comment]
@nadavbra,
I was able to reproduce the error with [TF v2.0](https://colab.research.google.com/gist/amahendrakar/1b34de0904e872a39ec7d02f4177b1cf/45966-2-0.ipynb). 

However, the error seems to be fixed with the latest stable version [v2.4](https://colab.research.google.com/gist/amahendrakar/595533f687d7b7a0a60895e5e30280a1/45966.ipynb). I was able to run the code without any issues with it. Please check the linked gist for reference. Thanks!
